cc/algorithms/partition-selection-strategies.h

```diff
@@ -0,0 +1,345 @@
+
+// Provides a common abstraction for PartitionSelectionStrategy.  Partition
+// selection strategies decide whether or not a specific partition should be
+// dropped or kept.
+class PartitionSelectionStrategy {
+ public:
+  //Builder base class
+  class Builder {
+   public:
+     virtual ~Builder() = default;
+
+     Builder& SetEpsilon(double epsilon) {
+       epsilon_ = epsilon;
+       return *static_cast<Builder*>(this);
+     }
+
+     Builder& SetDelta(double delta) {
+       delta_ = delta;
+       return *static_cast<Builder*>(this);
+     }
+
+     Builder& SetMaxPartitionsContributed(int max_partitions_contributed) {
+       max_partitions_contributed_ = max_partitions_contributed;
+       return *static_cast<Builder*>(this);
+     }
+
+     virtual base::StatusOr<std::unique_ptr<PartitionSelectionStrategy>>
+     Build() = 0;
+
+   protected:
+     absl::optional<double> epsilon_;
+     absl::optional<double> delta_;
+     absl::optional<int> max_partitions_contributed_;
+
+     //Checks if the max number of partitions contributed  to is set and valid.
+     base::Status MaxPartitionsContributedIsSetAndValid() {
+       if (!max_partitions_contributed_.has_value()) {
+         return base::InvalidArgumentError(
+          "Max number of partitions a user can contribute to has to be set.");
+       }
+       if (max_partitions_contributed_ <= 0) {
+         return base::InvalidArgumentError(absl::StrCat(
+           "Max number of partitions a user can contribute to has to be"
+           " positive but is ",
+           delta_.value()));
+       }
+       return base::OkStatus();
+     }
+
+     // Checks if delta is set and valid.
+     base::Status DeltaIsSetAndValid() {
+       if (!delta_.has_value()) {
+         return base::InvalidArgumentError("Delta has to be set.");
+       }
+       if (!std::isfinite(delta_.value())) {
+         return base::InvalidArgumentError(
+           absl::StrCat("Delta has to be finite but is ", delta_.value()));
+       }
+       if (delta_.value() <= 0 || 1 <= delta_.value()) {
+         return base::InvalidArgumentError(absl::StrCat(
+           "Delta has to be in the interval (0,1) but is ",
+           delta_.value()));
+        }
+        return base::OkStatus();
+      }
+  };
+
+  PartitionSelectionStrategy(double epsilon, double delta,
+                             int max_partitions_contributed)
+  	: epsilon_(epsilon), delta_(delta),
+      adjusted_delta_(1.0 - pow(1 - delta,
+                      (double) max_partitions_contributed)),
+      max_partitions_contributed_(max_partitions_contributed) {}
+
+  virtual ~PartitionSelectionStrategy() = default;
+
+  double GetEpsilon() { return epsilon_; }
+
+  double GetDelta() { return delta_; }
+
+  double GetAdjustedDelta() { return adjusted_delta_; }
+
+  double GetMaxPartitionsContributed() { return max_partitions_contributed_; }
+
+  virtual bool ShouldKeep(int num_users) = 0;
+
+ protected:
+  double epsilon_;
```

Comment:
Uh oh... it doesn't work as private.  I think it's because the child classes need to access these variables as well.  So I can't make this change in the base class, but I can do it for the child classes.

Reply:
The style guide's suggestion is to make it private with a protected getter (or setter, as necessary).